### PR TITLE
CPP: Fix performance issues in ClassesWithManyFields.ql

### PR DESCRIPTION
**Pull request by geoffw0**
As noted in the weekly meeting, this query performed extremely poorly (timed out on my machine) on Chromium due to it's complex logic for blaming particular groups of member variables in the violation message.  It turns out this logic was full of cartesian products on the number of variable declarations on one line or nearby - a problem that is particularly pronounced on Chrome (but most projects exhibit a lesser version of this issue e.g. in template / macro logic or line 0 of file `""`).

The new version makes use of modern QL features to make an abstraction for a line containing one or more variable declaration entries - quickly forgetting about those individual VDEs themselves, resulting in good performance.

### CODE REVIEW

File: cpp/ql/src/Architecture/Refactoring Opportunities/ClassesWithManyFields.ql [hunk 1]
```diff
@@ -20,66 +23,143 @@ string kindstr(Class c)
...
+
+  /**
+   * Gets the end column of the last `VariableDeclarationEntry` on this line.
+   */
+  int getEndColumn() {
+    result = max(getAVDE().getLocation().getEndColumn())
+  }
+
+  /**
+   * Gets the rank of this `VariableDeclarationLine` in it's file and class
```
Comment (dave-bartolomeo) [inline cpp/ql/src/Architecture/Refactoring Opportunities/ClassesWithManyFields.ql]:
```suggestion
   * Gets the rank of this `VariableDeclarationLine` in its file and class
```

**Comment by geoffw0** [comment]
Good spot, I've accepted your suggestion.

File: cpp/ql/test/query-tests/Architecture/Refactoring Opportunities/ClassesWithManyFields/cwmf.expected [hunk 2]
```diff
@@ -1,7 +1,7 @@
...
+| cwmf.cpp:13:3:14:12 | bb | Class bb has 20 fields; we suggest refactoring to 15 fields or fewer. | cwmf.cpp:13:3:14:12 | group of 20 fields here | group of 20 fields here |
+| cwmf.cpp:24:3:25:12 | dd<T> | Template class dd<T> has 20 fields; we suggest refactoring to 15 fields or fewer. | cwmf.cpp:24:3:25:12 | group of 20 fields here | group of 20 fields here |
+| cwmf.cpp:30:3:31:12 | ee<U> | Template class ee<U> has 20 fields; we suggest refactoring to 15 fields or fewer. | cwmf.cpp:30:3:31:12 | group of 20 fields here | group of 20 fields here |
+| cwmf.cpp:41:8:57:22 | MyParticle | Struct MyParticle has 30 fields; we suggest refactoring to 15 fields or fewer. | cwmf.cpp:41:8:57:22 | group of 30 fields here | group of 30 fields here |
+| different_types.h:15:15:33:10 | DifferentTypes2 | Class DifferentTypes2 has 18 fields; we suggest refactoring to 15 fields or fewer. | different_types.h:15:15:33:10 | group of 18 fields here | group of 18 fields here |
+| different_types.h:15:15:33:10 | DifferentTypes2 | Class DifferentTypes2 has 18 fields; we suggest refactoring to 15 fields or fewer. | different_types.h:15:15:33:10 | group of 18 fields here | group of 18 fields here |
```
Comment (jbj) [inline cpp/ql/test/query-tests/Architecture/Refactoring Opportunities/ClassesWithManyFields/cwmf.expected]: It looks like there are no tests with `suffix != ""`. Please add a test or two like that so we can see that your changes work for that case.

File: cpp/ql/src/Architecture/Refactoring Opportunities/ClassesWithManyFields.ql [hunk 3]
```diff
@@ -20,66 +23,143 @@ string kindstr(Class c)
...
+  Class getClass() {
+    result = c
+  }
+
+  /**
+   * Gets the line of this `VariableDeclarationLine`.
+   */
+  int getLine() {
+    result = line
+  }
```
Comment (jbj) [inline cpp/ql/src/Architecture/Refactoring Opportunities/ClassesWithManyFields.ql]: Please auto-format this query. I'm guessing that'll be necessary in any case to deal with the merge conflicts caused by #1239.

File: cpp/ql/test/query-tests/Architecture/Refactoring Opportunities/ClassesWithManyFields/cwmf.expected (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/test/query-tests/Architecture/Refactoring Opportunities/ClassesWithManyFields/cwmf.expected]: Done.

File: cpp/ql/src/Architecture/Refactoring Opportunities/ClassesWithManyFields.ql (hunk 3 above)
Comment (geoffw0) [reply cpp/ql/src/Architecture/Refactoring Opportunities/ClassesWithManyFields.ql]: Done.

File: cpp/ql/src/Architecture/Refactoring Opportunities/ClassesWithManyFields.ql (hunk 1 above)
Comment (Hinaroos67) [reply cpp/ql/src/Architecture/Refactoring Opportunities/ClassesWithManyFields.ql]: 331cc497e6f30985e616983a5ee790de74b8554f

**Pull request merged**